File: examples/clients/javascript/songbird-client.js
/**
 * Songbird JSON-RPC JavaScript/Node.js Client
 * Universal Gateway Client for Songbird Orchestrator
 * 
 * Usage (Node.js):
 *   const { SongbirdClient } = require('./songbird-client');
 *   const client = new SongbirdClient('http://localhost:8080');
 *   const health = await client.health();
 *   console.log(health);
 * 
 * Usage (Browser):
 *   <script src="songbird-client.js"></script>
 *   <script>
 *     const client = new SongbirdClient('http://localhost:8080');
 *     client.health().then(console.log);
 *   </script>
 * 
 * Version: 0.2.1
 * Last Updated: November 11, 2025
 */

/**
 * Custom error class for Songbird client errors
 */
class SongbirdError extends Error {
    constructor(code, message, data = null) {
        super(`Error ${code}: ${message}`);
        this.name = 'SongbirdError';
        this.code = code;
        this.data = data;
    }
}

/**
 * Songbird JSON-RPC Client
 * 
 * Simple, universal client for Songbird's JSON-RPC API
 * Works in both Node.js and modern browsers
 * 
 * @example
 * const client = new SongbirdClient('http://localhost:8080');
 * const health = await client.health();
 * console.log(health.status); // 'healthy'
 */
class SongbirdClient {
    /**
     * Create a Songbird client
     * 
     * @param {string} baseUrl - Songbird orchestrator URL (default: http://localhost:8080)
     * @param {number} timeout - Request timeout in milliseconds (default: 30000)
     */
    constructor(baseUrl = 'http://localhost:8080', timeout = 30000) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.jsonrpcUrl = `${this.baseUrl}/jsonrpc`;
        this.timeout = timeout;
        this.requestId = 0;
    }

    /**
     * Get next request ID
     * @private
     */
    _getNextId() {
        return ++this.requestId;
    }

    /**
     * Make a JSON-RPC call
     * 
     * @private
     * @param {string} method - JSON-RPC method name
     * @param {object} [params] - Method parameters
     * @returns {Promise<any>} Method result
     * @throws {SongbirdError} If the call fails
     */
    async _call(method, params = null) {
        const payload = {
            jsonrpc: '2.0',
            method: method,
            id: this._getNextId()
        };

        if (params !== null) {
            payload.params = params;
        }

        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        try {
            const response = await fetch(this.jsonrpcUrl, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(payload),
                signal: controller.signal
            });

            clearTimeout(timeoutId);

            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }

            const data = await response.json();

            if (data.error) {
                throw new SongbirdError(
                    data.error.code || -1,
                    data.error.message || 'Unknown error',
                    data.error.data
                );
            }

            return data.result;
        } catch (error) {
            clearTimeout(timeoutId);
            if (error.name === 'AbortError') {
                throw new Error(`Request timeout after ${this.timeout}ms`);
            }
            throw error;
        }
    }

    /**
     * Make multiple JSON-RPC calls in one HTTP request
     * 
     * @private
     * @param {Array<{method: string, params?: object}>} requests - List of requests
     * @returns {Promise<Array<any>>} List of results (may include SongbirdError objects)
     */
    async _batchCall(requests) {
        const payload = requests.map(req => ({
            jsonrpc: '2.0',
            method: req.method,
            params: req.params || null,
            id: this._getNextId()
        }));

        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        try {
            const response = await fetch(this.jsonrpcUrl, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(payload),
                signal: controller.signal
            });

            clearTimeout(timeoutId);

            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }

            const results = await response.json();

            // Return results (converting errors to SongbirdError objects)
            return results.map(result => {
                if (result.error) {
                    return new SongbirdError(
                        result.error.code || -1,
                        result.error.message || 'Unknown error',
                        result.error.data
                    );
                }
                return result.result;
            });
        } catch (error) {
            clearTimeout(timeoutId);
            if (error.name === 'AbortError') {
                throw new Error(`Request timeout after ${this.timeout}ms`);
            }
            throw error;
        }
    }

    // ============================================
    // Health & Info Methods
    // ============================================

    /**
     * Check Songbird health
     * 
     * @returns {Promise<{status: string, version: string, uptime_seconds: number}>}
     * 
     * @example
     * const health = await client.health();
     * console.log(health.status); // 'healthy'
     */
    async health() {
        return this._call('songbird.health');
    }

    /**
     * Get Songbird version information
     * 
     * @returns {Promise<{version: string, name: string, commit?: string, build_date?: string}>}
     * 
     * @example
     * const version = await client.version();
     * console.log(version.version); // '0.2.1'
     */
    async version() {
        return this._call('songbird.version');
    }

    // ============================================
    // Protocol Methods
    // ============================================

    /**
     * Get available protocol capabilities
     * 
     * @returns {Promise<{protocols: object}>}
     * 
     * @example
     * const caps = await client.protocolCapabilities();
     * console.log(Object.keys(caps.protocols)); // ['http', 'jsonrpc', ...]
     */
    async protocolCapabilities() {
        return this._call('songbird.protocol.capabilities');
    }

    // ============================================
    // Service Management Methods
    // ============================================

    /**
     * List all registered services
     * 
     * @returns {Promise<Array<object>>}
     * 
     * @example
     * const services = await client.listServices();
     * services.forEach(s => console.log(`${s.name}: ${s.endpoint}`));
     */
    async listServices() {
        const result = await this._call('songbird.services.list');
        return result.services || [];
    }

    /**
     * Get details for a specific service
     * 
     * @param {string} serviceId - Service identifier
     * @returns {Promise<object>}
     * 
     * @example
     * const service = await client.getService('my-service-123');
     * console.log(service.endpoint);
     */
    async getService(serviceId) {
        return this._call('songbird.services.get', { service_id: serviceId });
    }

    /**
     * Register a new service with Songbird
     * 
     * @param {object} serviceConfig - Service configuration
     * @returns {Promise<object>}
     * 
     * @example
     * const result = await client.registerService({
     *     name: 'my-service',
     *     endpoint: 'http://localhost:3000',
     *     capabilities: ['compute', 'storage']
     * });
     */
    async registerService(serviceConfig) {
        return this._call('songbird.services.register', serviceConfig);
    }

    // ============================================
    // Compute Methods
    // ============================================

    /**
     * Schedule a compute task (integrates with Toadstool)
     * 
     * @param {object} taskConfig - Task configuration
     * @returns {Promise<{task_id: string}>}
     * 
     * @example
     * const result = await client.scheduleCompute({
     *     task: 'train_model',
     *     language: 'python',
     *     code: 'import numpy as np\nprint("Hello!")'
     * });
     * console.log(result.task_id);
     */
    async scheduleCompute(taskConfig) {
        return this._call('songbird.compute.schedule', taskConfig);
    }

    /**
     * Get status of a compute task
     * 
     * @param {string} taskId - Task identifier
     * @returns {Promise<{status: string, progress?: number, result?: any}>}
     * 
     * @example
     * const status = await client.getComputeStatus('task-123');
     * console.log(status.status); // 'completed', 'running', 'failed'
     */
    async getComputeStatus(taskId) {
        return this._call('songbird.compute.status', { task_id: taskId });
    }

    /**
     * Wait for a compute task to complete
     * 
     * @param {string} taskId - Task identifier
     * @param {number} pollInterval - Seconds between status checks (default: 1.0)
     * @param {number|null} timeout - Maximum wait time in seconds (default: null = forever)
     * @returns {Promise<object>} Final task status
     * @throws {Error} If timeout is exceeded
     * 
     * @example
     * const task = await client.scheduleCompute({...});
     * const result = await client.waitForCompute(task.task_id, 1, 300);
     * console.log(result.result);
     */
    async waitForCompute(taskId, pollInterval = 1.0, timeout = null) {
        const startTime = Date.now();

        while (true) {
            const status = await this.getComputeStatus(taskId);

            if (['completed', 'failed', 'cancelled'].includes(status.status)) {
                return status;
            }

            if (timeout && (Date.now() - startTime) / 1000 > timeout) {
                throw new Error(`Task ${taskId} did not complete within ${timeout} seconds`);
            }

            await new Promise(resolve => setTimeout(resolve, pollInterval * 1000));
        }
    }

    // ============================================
    // Federation Methods
    // ============================================

    /**
     * List all federation peers
     * 
     * @returns {Promise<Array<object>>}
     * 
     * @example
     * const peers = await client.listFederationPeers();
     * peers.forEach(p => console.log(`${p.id}: ${p.endpoint}`));
     */
    async listFederationPeers() {
        const result = await this._call('songbird.federation.peers');
        return result.peers || [];
    }

    /**
     * Join a federation by connecting to a peer
     * 
     * @param {object} peerConfig - Peer configuration
     * @returns {Promise<object>}
     * 
     * @example
     * const result = await client.joinFederation({
     *     peer_id: 'tower-a',
     *     endpoint: 'http://tower-a.example.com:8080'
     * });
     */
    async joinFederation(peerConfig) {
        return this._call('songbird.federation.join', peerConfig);
    }

    // ============================================
    // Batch Operations
    // ============================================

    /**
     * Execute multiple operations in one HTTP request
     * 
     * @param {Array<{method: string, params?: object}>} operations - List of operations
     * @returns {Promise<Array<any>>} List of results
     * 
     * @example
     * const [health, version, services] = await client.batch([
     *     { method: 'songbird.health' },
     *     { method: 'songbird.version' },
     *     { method: 'songbird.services.list' }
     * ]);
     */
    async batch(operations) {
        return this._batchCall(operations);
    }

    // ============================================
    // Utility Methods
    // ============================================

    /**
     * Quick health check
     * 
     * @returns {Promise<boolean>} True if healthy, false otherwise
     * 
     * @example
     * if (await client.isHealthy()) {
     *     console.log('Songbird is up!');
     * }
     */
    async isHealthy() {
        try {
            const health = await this.health();
            return health.status === 'healthy';
        } catch (error) {
            return false;
        }
    }
}

// ============================================
// Exports
// ============================================

// Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { SongbirdClient, SongbirdError };
}

// Browser
if (typeof window !== 'undefined') {
    window.SongbirdClient = SongbirdClient;
    window.SongbirdError = SongbirdError;
}

// ============================================
// Example Usage (Node.js)
// ============================================

if (typeof require !== 'undefined' && require.main === module) {
    (async () => {
        const client = new SongbirdClient('http://localhost:8080');

        // Example 1: Basic health check
        console.log('='.repeat(60));
        console.log('Example 1: Basic Health Check');
        console.log('='.repeat(60));

        try {
            const health = await client.health();
            console.log(`✅ Songbird is ${health.status}`);
            console.log(`   Version: ${health.version}`);
            console.log(`   Uptime: ${health.uptime_seconds} seconds`);
        } catch (error) {
            console.log(`❌ Failed to connect: ${error.message}`);
        }

        console.log();

        // Example 2: Get version info
        console.log('='.repeat(60));
        console.log('Example 2: Version Information');
        console.log('='.repeat(60));

        try {
            const version = await client.version();
            console.log(`Name: ${version.name}`);
            console.log(`Version: ${version.version}`);
            console.log(`Commit: ${version.commit || 'N/A'}`);
            console.log(`Build Date: ${version.build_date || 'N/A'}`);
        } catch (error) {
            console.log(`❌ Failed: ${error.message}`);
        }

        console.log();

        // Example 3: Protocol capabilities
        console.log('='.repeat(60));
        console.log('Example 3: Protocol Capabilities');
        console.log('='.repeat(60));

        try {
            const caps = await client.protocolCapabilities();
            const protocols = caps.protocols || {};
            console.log('Available Protocols:');
            for (const [name, info] of Object.entries(protocols)) {
                console.log(`  • ${name}: ${info.version || 'N/A'}`);
            }
        } catch (error) {
            console.log(`❌ Failed: ${error.message}`);
        }

        console.log();

        // Example 4: Batch operations
        console.log('='.repeat(60));
        console.log('Example 4: Batch Operations');
        console.log('='.repeat(60));

        try {
            const [health, version, caps] = await client.batch([
                { method: 'songbird.health' },
                { method: 'songbird.version' },
                { method: 'songbird.protocol.capabilities' }
            ]);

            console.log('✅ Executed 3 operations in 1 HTTP request');
            console.log(`   Health: ${health.status}`);
            console.log(`   Version: ${version.version}`);
            console.log(`   Protocols: ${Object.keys(caps.protocols).length} available`);
        } catch (error) {
            console.log(`❌ Failed: ${error.message}`);
        }

        console.log();

        // Example 5: Quick health check
        console.log('='.repeat(60));
        console.log('Example 5: Quick Health Check');
        console.log('='.repeat(60));

        if (await client.isHealthy()) {
            console.log('✅ Songbird is healthy and ready!');
        } else {
            console.log('❌ Songbird is not responding');
        }

        console.log();
        console.log('='.repeat(60));
        console.log('🎉 All examples complete!');
        console.log('='.repeat(60));
    })();
}

File: examples/clients/python/songbird_client.py
#!/usr/bin/env python3
"""
Songbird JSON-RPC Python Client
Universal Gateway Client for Songbird Orchestrator

Usage:
    from songbird_client import SongbirdClient
    
    client = SongbirdClient("http://localhost:8080")
    health = client.health()
    print(health)

Version: 0.2.1
Last Updated: April 16, 2026
"""

import requests
from typing import Any, Dict, List, Optional
import time


class SongbirdError(Exception):
    """Base exception for Songbird client errors"""
    def __init__(self, code: int, message: str, data: Any = None):
        self.code = code
        self.message = message
        self.data = data
        super().__init__(f"Error {code}: {message}")


class SongbirdClient:
    """
    Simple, universal client for Songbird JSON-RPC API
    
    Example:
        >>> client = SongbirdClient("http://localhost:8080")
        >>> health = client.health()
        >>> print(health['status'])
        'healthy'
    """
    
    def __init__(self, base_url: str = "http://localhost:8080", timeout: int = 30):
        """
        Initialize Songbird client
        
        Args:
            base_url: Songbird orchestrator URL (default: http://localhost:8080)
            timeout: Request timeout in seconds (default: 30)
        """
        self.base_url = base_url.rstrip('/')
        self.jsonrpc_url = f"{self.base_url}/jsonrpc"
        self.timeout = timeout
        self.session = requests.Session()
        self._request_id = 0
    
    def _get_next_id(self) -> int:
        """Get next request ID"""
        self._request_id += 1
        return self._request_id
    
    def _call(self, method: str, params: Optional[Dict[str, Any]] = None) -> Any:
        """
        Make a JSON-RPC call
        
        Args:
            method: JSON-RPC method name
            params: Method parameters (optional)
            
        Returns:
            Method result
            
        Raises:
            SongbirdError: If the call fails
            requests.RequestException: If HTTP request fails
        """
        payload = {
            "jsonrpc": "2.0",
            "method": method,
            "id": self._get_next_id()
        }
        
        if params is not None:
            payload["params"] = params
        
        response = self.session.post(
            self.jsonrpc_url,
            json=payload,
            timeout=self.timeout
        )
        response.raise_for_status()
        
        data = response.json()
        
        if "error" in data:
            error = data["error"]
            raise SongbirdError(
                code=error.get("code", -1),
                message=error.get("message", "Unknown error"),
                data=error.get("data")
            )
        
        return data.get("result")
    
    def _batch_call(self, requests_list: List[Dict[str, Any]]) -> List[Any]:
        """
        Make multiple JSON-RPC calls in one HTTP request
        
        Args:
            requests_list: List of request dicts with 'method' and optional 'params'
            
        Returns:
            List of results (in same order as requests)
        """
        payload = [
            {
                "jsonrpc": "2.0",
                "method": req["method"],
                "params": req.get("params"),
                "id": self._get_next_id()
            }
            for req in requests_list
        ]
        
        response = self.session.post(
            self.jsonrpc_url,
            json=payload,
            timeout=self.timeout
        )
        response.raise_for_status()
        
        results = response.json()
        
        # Return results in order
        ordered_results = []
        for result in results:
            if "error" in result:
                error = result["error"]
                ordered_results.append(SongbirdError(
                    code=error.get("code", -1),
                    message=error.get("message", "Unknown error"),
                    data=error.get("data")
                ))
            else:
                ordered_results.append(result.get("result"))
        
        return ordered_results
    
    # ============================================
    # Health & Info Methods
    # ============================================
    
    def health(self) -> Dict[str, Any]:
        """
        Check Songbird health
        
        Returns:
            Health status dict with 'status', 'version', 'uptime_seconds'
            
        Example:
            >>> client.health()
            {'status': 'healthy', 'version': '0.2.1', 'uptime_seconds': 1234}
        """
        return self._call("songbird.health")
    
    def version(self) -> Dict[str, Any]:
        """
        Get Songbird version information
        
        Returns:
            Version info dict with 'version', 'name', 'commit', 'build_date'
            
        Example:
            >>> client.version()
            {'version': '0.2.1', 'name': 'Songbird Universal Orchestrator', ...}
        """
        return self._call("songbird.version")
    
    # ============================================
    # Protocol Methods
    # ============================================
    
    def protocol_capabilities(self) -> Dict[str, Any]:
        """
        Get available protocol capabilities
        
        Returns:
            Protocol capabilities dict with 'protocols' listing available protocols
            
        Example:
            >>> caps = client.protocol_capabilities()
            >>> print(caps['protocols'].keys())
            dict_keys(['http', 'jsonrpc', 'tarpc', 'websocket'])
        """
        return self._call("songbird.protocol.capabilities")
    
    # ============================================
    # Service Management Methods
    # ============================================
    
    def list_services(self) -> List[Dict[str, Any]]:
        """
        List all registered services
        
        Returns:
            List of service dicts
            
        Example:
            >>> services = client.list_services()
            >>> for service in services:
            ...     print(f"{service['name']}: {service['endpoint']}")
        """
        result = self._call("songbird.services.list")
        return result.get("services", [])
    
    def get_service(self, service_id: str) -> Dict[str, Any]:
        """
        Get details for a specific service
        
        Args:
            service_id: Service identifier
            
        Returns:
            Service details dict
            
        Example:
            >>> service = client.get_service("my-service-123")
            >>> print(service['endpoint'])
        """
        return self._call("songbird.services.get", {"service_id": service_id})
    
    def register_service(self, service_config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Register a new service with Songbird
        
        Args:
            service_config: Service configuration dict
            
        Returns:
            Registration result
            
        Example:
            >>> result = client.register_service({
            ...     "name": "my-service",
            ...     "endpoint": "http://localhost:3000",
            ...     "capabilities": ["compute", "storage"]
            ... })
        """
        return self._call("songbird.services.register", service_config)
    
    # ============================================
    # Compute Methods
    # ============================================
    
    def schedule_compute(self, task_config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Schedule a compute task (integrates with Toadstool)
        
        Args:
            task_config: Task configuration dict with 'task', 'language', 'code', etc.
            
        Returns:
            Task scheduling result with 'task_id'
            
        Example:
            >>> result = client.schedule_compute({
            ...     "task": "train_model",
            ...     "language": "python",
            ...     "code": "import numpy as np\\nprint('Hello!')"
            ... })
            >>> task_id = result['task_id']
        """
        return self._call("songbird.compute.schedule", task_config)
    
    def get_compute_status(self, task_id: str) -> Dict[str, Any]:
        """
        Get status of a compute task
        
        Args:
            task_id: Task identifier from schedule_compute
            
        Returns:
            Task status dict with 'status', 'progress', 'result', etc.
            
        Example:
            >>> status = client.get_compute_status("task-123")
            >>> print(status['status'])
            'completed'
        """
        return self._call("songbird.compute.status", {"task_id": task_id})
    
    def wait_for_compute(self, task_id: str, poll_interval: float = 1.0, timeout: Optional[float] = None) -> Dict[str, Any]:
        """
        Wait for a compute task to complete
        
        Args:
            task_id: Task identifier
            poll_interval: Seconds between status checks (default: 1.0)
            timeout: Maximum wait time in seconds (default: None = wait forever)
            
        Returns:
            Final task status
            
        Raises:
            TimeoutError: If timeout is exceeded
            
        Example:
            >>> task = client.schedule_compute({...})
            >>> result = client.wait_for_compute(task['task_id'], timeout=300)
            >>> print(result['result'])
        """
        start_time = time.time()
        
        while True:
            status = self.get_compute_status(task_id)
            
            if status["status"] in ["completed", "failed", "cancelled"]:
                return status
            
            if timeout and (time.time() - start_time) > timeout:
                raise TimeoutError(f"Task {task_id} did not complete within {timeout} seconds")
            
            time.sleep(poll_interval)
    
    # ============================================
    # Federation Methods
    # ============================================
    
    def list_federation_peers(self) -> List[Dict[str, Any]]:
        """
        List all federation peers
        
        Returns:
            List of peer dicts
            
        Example:
            >>> peers = client.list_federation_peers()
            >>> for peer in peers:
            ...     print(f"{peer['id']}: {peer['endpoint']}")
        """
        result = self._call("songbird.federation.peers")
        return result.get("peers", [])
    
    def join_federation(self, peer_config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Join a federation by connecting to a peer
        
        Args:
            peer_config: Peer configuration dict with 'peer_id' and 'endpoint'
            
        Returns:
            Join result
            
        Example:
            >>> result = client.join_federation({
            ...     "peer_id": "tower-a",
            ...     "endpoint": "http://tower-a.example.com:8080"
            ... })
        """
        return self._call("songbird.federation.join", peer_config)
    
    # ============================================
    # Batch Operations
    # ============================================
    
    def batch(self, operations: List[Dict[str, Any]]) -> List[Any]:
        """
        Execute multiple operations in one HTTP request
        
        Args:
            operations: List of operation dicts with 'method' and optional 'params'
            
        Returns:
            List of results (may include SongbirdError objects for failed operations)
            
        Example:
            >>> results = client.batch([
            ...     {"method": "songbird.health"},
            ...     {"method": "songbird.version"},
            ...     {"method": "songbird.services.list"}
            ... ])
            >>> health, version, services = results
        """
        return self._batch_call(operations)
    
    # ============================================
    # Utility Methods
    # ============================================
    
    def is_healthy(self) -> bool:
        """
        Quick health check
        
        Returns:
            True if healthy, False otherwise
            
        Example:
            >>> if client.is_healthy():
            ...     print("Songbird is up!")
        """
        try:
            health = self.health()
            return health.get("status") == "healthy"
        except Exception:
            return False
    
    def close(self):
        """Close the HTTP session"""
        self.session.close()
    
    def __enter__(self):
        """Context manager entry"""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.close()


# ============================================
# Convenience Functions
# ============================================

def create_client(base_url: str = "http://localhost:8080", timeout: int = 30) -> SongbirdClient:
    """
    Create a Songbird client
    
    Args:
        base_url: Songbird orchestrator URL
        timeout: Request timeout in seconds
        
    Returns:
        SongbirdClient instance
    """
    return SongbirdClient(base_url, timeout)


# ============================================
# Example Usage
# ============================================

if __name__ == "__main__":
    # Example 1: Basic usage
    print("=" * 60)
    print("Example 1: Basic Health Check")
    print("=" * 60)
    
    client = SongbirdClient("http://localhost:8080")
    
    try:
        health = client.health()
        print(f"✅ Songbird is {health['status']}")
        print(f"   Version: {health['version']}")
        print(f"   Uptime: {health['uptime_seconds']} seconds")
    except Exception as e:
        print(f"❌ Failed to connect: {e}")
    
    print()
    
    # Example 2: Get version info
    print("=" * 60)
    print("Example 2: Version Information")
    print("=" * 60)
    
    try:
        version = client.version()
        print(f"Name: {version['name']}")
        print(f"Version: {version['version']}")
        print(f"Commit: {version.get('commit', 'N/A')}")
        print(f"Build Date: {version.get('build_date', 'N/A')}")
    except Exception as e:
        print(f"❌ Failed: {e}")
    
    print()
    
    # Example 3: Protocol capabilities
    print("=" * 60)
    print("Example 3: Protocol Capabilities")
    print("=" * 60)
    
    try:
        caps = client.protocol_capabilities()
        protocols = caps.get("protocols", {})
        print(f"Available Protocols:")
        for proto_name, proto_info in protocols.items():
            print(f"  • {proto_name}: {proto_info.get('version', 'N/A')}")
    except Exception as e:
        print(f"❌ Failed: {e}")
    
    print()
    
    # Example 4: Batch operations
    print("=" * 60)
    print("Example 4: Batch Operations")
    print("=" * 60)
    
    try:
        results = client.batch([
            {"method": "songbird.health"},
            {"method": "songbird.version"},
            {"method": "songbird.protocol.capabilities"}
        ])
        
        health, version, caps = results
        print(f"✅ Executed 3 operations in 1 HTTP request")
        print(f"   Health: {health['status']}")
        print(f"   Version: {version['version']}")
        print(f"   Protocols: {len(caps['protocols'])} available")
    except Exception as e:
        print(f"❌ Failed: {e}")
    
    print()
    
    # Example 5: Context manager
    print("=" * 60)
    print("Example 5: Using Context Manager")
    print("=" * 60)
    
    with SongbirdClient("http://localhost:8080") as client:
        if client.is_healthy():
            print("✅ Songbird is healthy and ready!")
        else:
            print("❌ Songbird is not responding")
    
    print()
    print("=" * 60)
    print("🎉 All examples complete!")
    print("=" * 60)

